// eval5/dist/cjs/interpreter/messages.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Messages = exports.InterruptThrowReferenceError = exports.InterruptThrowSyntaxError = exports.InterruptThrowError = exports.ThrowTypeError = exports.ThrowReferenceError = exports.ThrowSyntaxError = exports.ThrowError = void 0;
var ThrowError = /** @class */ (function (_super) {
    __extends(ThrowError, _super);
    function ThrowError() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ThrowError;
}(Error));
exports.ThrowError = ThrowError;
var ThrowSyntaxError = /** @class */ (function (_super) {
    __extends(ThrowSyntaxError, _super);
    function ThrowSyntaxError() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ThrowSyntaxError;
}(SyntaxError));
exports.ThrowSyntaxError = ThrowSyntaxError;
var ThrowReferenceError = /** @class */ (function (_super) {
    __extends(ThrowReferenceError, _super);
    function ThrowReferenceError() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ThrowReferenceError;
}(ReferenceError));
exports.ThrowReferenceError = ThrowReferenceError;
var ThrowTypeError = /** @class */ (function (_super) {
    __extends(ThrowTypeError, _super);
    function ThrowTypeError() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ThrowTypeError;
}(TypeError));
exports.ThrowTypeError = ThrowTypeError;
var InterruptThrowError = /** @class */ (function (_super) {
    __extends(InterruptThrowError, _super);
    function InterruptThrowError() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return InterruptThrowError;
}(ThrowError));
exports.InterruptThrowError = InterruptThrowError;
var InterruptThrowSyntaxError = /** @class */ (function (_super) {
    __extends(InterruptThrowSyntaxError, _super);
    function InterruptThrowSyntaxError() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return InterruptThrowSyntaxError;
}(ThrowSyntaxError));
exports.InterruptThrowSyntaxError = InterruptThrowSyntaxError;
var InterruptThrowReferenceError = /** @class */ (function (_super) {
    __extends(InterruptThrowReferenceError, _super);
    function InterruptThrowReferenceError() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return InterruptThrowReferenceError;
}(ThrowReferenceError));
exports.InterruptThrowReferenceError = InterruptThrowReferenceError;
exports.Messages = {
    UnknownError: [3001, "%0", InterruptThrowError],
    ExecutionTimeOutError: [3002, "Script execution timed out after %0ms", InterruptThrowError],
    NodeTypeSyntaxError: [1001, "Unknown node type: %0", InterruptThrowReferenceError],
    BinaryOperatorSyntaxError: [1002, "Unknown binary operator: %0", InterruptThrowReferenceError],
    LogicalOperatorSyntaxError: [
        1003,
        "Unknown logical operator: %0",
        InterruptThrowReferenceError,
    ],
    UnaryOperatorSyntaxError: [1004, "Unknown unary operator: %0", InterruptThrowReferenceError],
    UpdateOperatorSyntaxError: [1005, "Unknown update operator: %0", InterruptThrowReferenceError],
    ObjectStructureSyntaxError: [
        1006,
        "Unknown object structure: %0",
        InterruptThrowReferenceError,
    ],
    AssignmentExpressionSyntaxError: [
        1007,
        "Unknown assignment expression: %0",
        InterruptThrowReferenceError,
    ],
    VariableTypeSyntaxError: [1008, "Unknown variable type: %0", InterruptThrowReferenceError],
    ParamTypeSyntaxError: [1009, "Unknown param type: %0", InterruptThrowReferenceError],
    AssignmentTypeSyntaxError: [1010, "Unknown assignment type: %0", InterruptThrowReferenceError],
    FunctionUndefinedReferenceError: [2001, "%0 is not a function", ThrowReferenceError],
    VariableUndefinedReferenceError: [2002, "%0 is not defined", ThrowReferenceError],
    IsNotConstructor: [2003, "%0 is not a constructor", ThrowTypeError],
};
